src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,173 @@
+        if (index >= size || index < 0) {
+            throw new IndexOutOfBoundsException(INVALID_INDEX_ERROR_MESSAGE);
+        }
+        T pastValue = getNode(index).value;
+        getNode(index).value = value;
+        return pastValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        if (index >= size || index < 0) {
+            throw new IndexOutOfBoundsException(INVALID_INDEX_ERROR_MESSAGE);
+        }
+        Node<T> node = getNode(index);
+        unlink(node);
+        --size;
+        return node.value;
     }
 
     @Override
     public boolean remove(T object) {
+        for (int i = 0; i < size; i++) {
+            T value = get(i);
+            if (object != null && object.equals(value) || object == value) {
+                remove(i);
+                return true;
+            }
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return tail == null;
+    }
+
+    private Node<T> findNodeFromHead(int index) {
+        int currentIndex = 0;
+        Node<T> node = head;
+
+        while (currentIndex++ != index) {
+            node = node.next;
+        }
+        return node;
+    }
+
+    private Node<T> findNodeFromTail(int index) {
+        int currentIndex = size - 1;
+        Node<T> node = tail;
+
+        while (currentIndex-- != index) {
+            node = node.prev;
+        }
+        return node;
+    }
+
+    private Node<T> getNode(int index) {
+        return (index > size / 2) ? findNodeFromTail(index) : findNodeFromHead(index);
+    }
+
+    private void link(Node<T> destNode, T value) {
+        Node<T> newNode = new Node<>(destNode.prev, value, destNode);
+        if (head == destNode) {
+            head = newNode;
+        } else {
+            destNode.prev.next = newNode;
+        }
+        destNode.prev = newNode;
+    }
+
+    private void unlink(Node<T> node) {
+        if (head == node) {
+            head = node.next;
+            if (node.next != null) {
+                node.next.prev = node.prev;
+            } else {
+                tail = node.prev;
+            }
+        } else if (tail == node) {
+            tail = node.prev;
+            node.prev.next = node.next;
+        } else {
+            node.prev.next = node.next;
+            node.next.prev = node.prev;
+        }
+    }
+
+    private void addFirstNode(T value) {
+        head = new Node<>(null, value, null);
+        tail = head;
+    }
+
+    private void addLastNode(T value) {
+        tail = new Node<>(tail, value, null);
+        tail.prev.next = tail;
+        if (head.next == null) {
+            head.next = tail;
+        }
+    }
+
+    private static class Node<T> {
+        private Node<T> prev;
+        private T value;
+        private Node<T> next;
+
+        private Node(Node<T> prev, T value, Node<T> next) {
```

Comment:
Don't use getters and setters in class Node. We have access to private fields of the inner class, so using getters or setters is redundant.